File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/AnyTrueEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import java.util.Iterator;
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;

/*
AnyTrue(argument List<Boolean>) Boolean

The AnyTrue operator returns true if any non-null element in the source is true.
If the source contains no non-null elements, false is returned.
If the source is null, the result is null.
*/

public class AnyTrueEvaluator {

    public static Boolean anyTrue(Object src) {
        if (src == null) {
            return false;
        }

        if (src instanceof Iterable) {
            Iterable<?> element = (Iterable<?>) src;
            Iterator<?> elemsItr = element.iterator();

            if (!elemsItr.hasNext()) { // empty list
                return false;
            }

            while (elemsItr.hasNext()) {
                Object exp = elemsItr.next();

                if (exp == null) { // skip null
                    continue;
                }

                if (exp instanceof Boolean) {
                    Boolean boolVal = (Boolean) exp;

                    if (Boolean.TRUE == boolVal) {
                        return true;
                    }
                } else {
                    throw new InvalidOperatorArgument(
                            "AnyTrue(List<Boolean>)",
                            String.format("AnyTrue(List<%s>)", exp.getClass().getName()));
                }
            }

            return false; // all null or all false
        }

        throw new InvalidOperatorArgument(
                "AnyTrue(List<Boolean>)",
                String.format("AnyTrue(%s)", src.getClass().getName()));
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/AnyTrueEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.Boolean
import org.opencds.cqf.cql.engine.runtime.CqlType
import org.opencds.cqf.cql.engine.runtime.List

/*
AnyTrue(argument List<Boolean>) Boolean

The AnyTrue operator returns true if any non-null element in the source is true.
If the source contains no non-null elements, false is returned.
If the source is null, the result is null.
*/
object AnyTrueEvaluator {
    @JvmStatic
    fun anyTrue(src: CqlType?): Boolean {
        if (src == null) {
            return Boolean.FALSE
        }

        if (src is List) {
            val element = src
            val elemsItr = element.iterator()

            if (!elemsItr.hasNext()) { // empty list
                return Boolean.FALSE
            }

            while (elemsItr.hasNext()) {
                val exp = elemsItr.next()

                if (exp == null) { // skip null
                    continue
                }

                if (exp is Boolean) {
                    val boolVal = exp

                    if (boolVal.value) {
                        return Boolean.TRUE
                    }
                } else {
                    throw InvalidOperatorArgument(
                        "AnyTrue(List<Boolean>)",
                        "AnyTrue(List<${exp.typeAsString}>)",
                    )
                }
            }

            return Boolean.FALSE // all null or all false
        }

        throw InvalidOperatorArgument("AnyTrue(List<Boolean>)", "AnyTrue(${src.typeAsString})")
    }
}
